src/index.ts

```diff
@@ -66,88 +79,312 @@ export default class QuickbooksConnector extends BaseHttpConnector<
+    }) // can be an array of multiple scopes ex : {scope:[OAuthClient.scopes.Accounting,OAuthClient.scopes.OpenId]
 
-    console.log(authUri)
+    console.log(`===================== authUri =====================\n
+    ${authUri}
+    \n===================================================`)
     return oauthClient
   }
 
+  async isInStore() {
+    const dbToken = await this.getQBToken()
+    if (dbToken) {
+      await this.setClient(dbToken.token)
+      return true
+    }
+    return false
+  }
+
+  async refreshTokenIfNeeded() {
+    let dbToken = await this.getQBToken()
+    if (!dbToken) {
+      console.error(`refreshTokenIfNeeded, stored token not found for ${this.options.realmId}`)
+      return
+    }
+    // this.oauthClient.setToken(dbToken.token)
+    const expiry = Number(dbToken.token.createdAt) + Number(dbToken.token.expires_in) * 1000
+    const needRefresh = (expiry - 120 * 1000) < Date.now() // 2 minutes before token is expired
+
+    if (needRefresh) {
+      console.log('Before refreshing', dbToken.token)
+      try {
+        this.oauthClient.setToken(dbToken.token)
+        const authResponse = await this.oauthClient.refresh()
+        await this.storeTokenAndSetClient(authResponse)
+        console.log('Token is refreshed')
+      }
+      catch (e: any) {
+        console.error('Refresh Token Error:', e)
+        console.error(e.intuit_tid)
+      }
+    }
+    return this.client
+  }
+
   // Use to exchange code for token
   async handle(req: Request, res: Response): Promise<boolean> {
-    console.log(req.query.code)
-    if (req.query.realmId) {
-      this.oauthClient.createToken(req.url)
-        .then(function(authResponse:any) {
-          console.log('The Token is  '+ JSON.stringify(authResponse.getJson()));
-        })
-        .catch(function(e:any) {
-          console.error("The error message is :"+e.originalMessage);
-          console.error(e.intuit_tid);
-        });
-        const {state, realmId, code} = req.query
-        if (typeof realmId === "string") {
-          this.options.realmId = parseInt(realmId, 10)
-        }
+    if (req.route.path === this.webhookPath) {
+      this.handleWebhook(req, res)
+    } else if (req.query.realmId) { // oauth callback
+      try {
+        const authResponse = await this.oauthClient.createToken(req.url)
+        await this.storeTokenAndSetClient(authResponse)
+      }
+      catch (e: any) {
+        console.error('The error message is :',e)
+        console.error(e.intuit_tid)
+      }
+      /* Do we need this
+      const { state, realmId, code } = req.query
+      if (typeof realmId === "string") {
+        this.options.realmId = realmId
+      }*/
     } else {
-      res.send({'text':'Error'})
+      res.send({ 'text': 'Error' })
     }
     res.sendStatus(200)
     return true
   }
+
+  private async handleWebhook(req: Request, res: Response) {
```

Comment:
the Reshuffle app calls handle, not sure how it works with a different name here